core/src/smartcontracts/isi/triggers/set.rs

```diff
@@ -633,52 +654,96 @@ impl Set {
-                .remove(id)
-                .map(extract_blob_hash)
-                .expect("`Set::time_triggers` doesn't contain required id. This is a bug"),
-            TriggeringEventType::ExecuteTrigger => self
-                .by_call_triggers
-                .remove(id)
-                .map(extract_blob_hash)
-                .expect("`Set::by_call_triggers` doesn't contain required id. This is a bug"),
+        let removed = match event_type {
+            TriggeringEventType::Data => {
+                Self::remove_from(&mut self.original_contracts, &mut self.data_triggers, id)
+            }
+            TriggeringEventType::Pipeline => Self::remove_from(
+                &mut self.original_contracts,
+                &mut self.pipeline_triggers,
+                id,
+            ),
+            TriggeringEventType::Time => {
+                Self::remove_from(&mut self.original_contracts, &mut self.time_triggers, id)
+            }
+            TriggeringEventType::ExecuteTrigger => {
+                Self::remove_from(&mut self.original_contracts, &mut self.by_call_triggers, id)
+            }
         };
 
-        if let Some(blob_hash) = blob_hash {
-            self.original_contracts
-                .remove(&blob_hash)
-                .expect("`Set::original_contracts` doesn't contain required hash. This is a bug");
-        }
+        assert!(
+            removed,
+            "`Set`'s `ids` and typed trigger collections are inconsistent. This is a bug"
+        );
 
         true
     }
 
+    /// Remove trigger from `triggers` and decrease the counter of the original [`WasmSmartContract`].
+    ///
+    /// Note that this function doesn't remove the trigger from [`Set::ids`].
+    ///
+    /// Returns `true` if trigger was removed and `false` otherwise.
+    fn remove_from<F: Filter>(
+        original_contracts: &mut IndexMap<
+            HashOf<WasmSmartContract>,
+            (WasmSmartContract, NonZeroU64),
+        >,
+        triggers: &mut IndexMap<TriggerId, LoadedAction<F>>,
+        trigger_id: &TriggerId,
+    ) -> bool {
+        triggers
+            .remove(trigger_id)
+            .map(|loaded_action| {
+                if let Some(blob_hash) = Self::extract_blob_hash(loaded_action) {
+                    Self::remove_original_trigger(original_contracts, blob_hash);
+                }
+            })
+            .is_some()
+    }
+
+    /// Decrease the counter of the original [`WasmSmartContract`] by `blob_hash`
+    /// or remove it if the counter reaches zero.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `blob_hash` is not in the [`Set::original_contracts`].
+    fn remove_original_trigger(
+        original_contracts: &mut IndexMap<
+            HashOf<WasmSmartContract>,
+            (WasmSmartContract, NonZeroU64),
+        >,
+        blob_hash: HashOf<WasmSmartContract>,
+    ) {
+        (|| {
+            let count = &mut original_contracts.get_mut(&blob_hash)?.1;
```

Comment:
I think using `.entry` might be nicer here? By pattern matching on it, you would have only one case to raise the "`Set::original_contracts` doesn't contain required hash" panic without having to use 

Something like:

```rust
match original_contracts.entry(blob_hash) {
  Occupied(o) => {
    let (_, count) = o.get(); // I think get_mut won't work here without polonius
    if ... {
      *o.get_mut().1 = new_count;
    } else {
      o.remove_entry();
    }
  }
  Vacant(_) => panic!("..."),
}
```

Reply:
You're right. Currently I'm doing the search twice. Don't know how I missed that